### ISSUE TITLE
simp only is quadratic in the size of the output term

### ISSUE THREAD

**Issue by JasonGross** [leanprover-community/lean#401]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover-community/lean/issues).
        * Specifically, check out the [wishlist](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3AI-wishlist), open [RFCs](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3ARFC),
        or [feature requests](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3AFeature).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

```lean
universes u v w ℓ

axiom f : ℕ → ℕ
axiom g : ℕ → ℕ

@[simp]
def comp_pow {A : Type u} (f : A → A) : ℕ → A → A
| 0 x := x
| (nat.succ n) x := f (comp_pow n x)

@[simp] def goal100 : Σ' P, P = (∀ x, comp_pow f 100 x = comp_pow g 100 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal200 : Σ' P, P = (∀ x, comp_pow f 200 x = comp_pow g 200 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal300 : Σ' P, P = (∀ x, comp_pow f 300 x = comp_pow g 300 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal400 : Σ' P, P = (∀ x, comp_pow f 400 x = comp_pow g 400 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal500 : Σ' P, P = (∀ x, comp_pow f 500 x = comp_pow g 500 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal600 : Σ' P, P = (∀ x, comp_pow f 600 x = comp_pow g 600 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal700 : Σ' P, P = (∀ x, comp_pow f 700 x = comp_pow g 700 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal800 : Σ' P, P = (∀ x, comp_pow f 800 x = comp_pow g 800 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal900 : Σ' P, P = (∀ x, comp_pow f 900 x = comp_pow g 900 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

@[simp] def goal1000 : Σ' P, P = (∀ x, comp_pow f 1000 x = comp_pow g 1000 x) :=
by { eapply psigma.mk, simp only [comp_pow] }

```

running `lean --profile` (Lean (version 3.4.2, commit cbd2b6686ddb, Release)) on this gives the times

n | time (s)
--|--
100 | 0.033
200 | 0.055
300 | 0.081
400 | 0.115
500 | 0.153
600 | 0.193
700 | 0.251
800 | 0.356
900 | 0.358
1000 | 0.46

which is quadratic:

![time of simp only vs size (1)](https://user-images.githubusercontent.com/396076/88102725-08a39200-cb6e-11ea-8c59-8158502dea71.png)

### Steps to Reproduce

See above

**Expected behavior:** Linear performance of reduction

**Actual behavior:** Quadratic performance of reduction

**Reproduces how often:** 100% of the time

### Versions

Lean (version 3.4.2, commit cbd2b6686ddb, Release)
```
$ uname -a
Linux jgross-Leopard-WS 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.4 LTS
Release:        18.04
Codename:       bionic
```

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by cipher1024** [comment]
I don't think this should be considered a bug. The term is traversed completely at every pass of `simp` and every pass meets only one opportunity for rewrite. I pointed out one way in https://github.com/leanprover-community/mathlib/issues/3500#issuecomment-662087225 to reduce the number of passes. Another possibility is to increase the number of opportunity to rewrite at each pass with lemmas of this shape

```lean
lemma comp_pow_bit0 : comp_pow f (bit0 n) x = comp_pow f n (comp_pow f n x) := ...
```

The number of passes should be logarithmic and the amount of work should be roughly linear. That's no my favorite solution for this specific problem but it's something worth considering when speeding up your rewrites.

**Comment by dselsam** [comment]
`simp` recurses a linear number of times on these examples. It is probably a cache/collision issue. If you prove the theorems in reverse order, it looks linear:
```
elaboration of goal1000 took 296ms
elaboration of goal900 took 319ms
elaboration of goal800 took 270ms
elaboration of goal700 took 216ms
elaboration of goal600 took 167ms
elaboration of goal500 took 130ms
elaboration of goal400 took 103ms
elaboration of goal300 took 76.8ms
elaboration of goal200 took 50.4ms
elaboration of goal100 took 28.8ms
```

**Comment by dselsam** [comment]
`simp` recurses a linear number of times on these examples, and the time within the `simp` tactic seems linear even out to 4000:
```
[4000]   1628ms    46.2%   tactic.simplify
[2000]    830ms    59.6%   tactic.simplify
```
It is the `apply` following `simp` that blows up:
```
[4000]  516ms    37.0%   tactic.apply_core
[2000]  114ms    27.5%   tactic.apply_core
```
Here `apply` is called by `relation_tactic` following simplification. It successfully applies `eq.refl`, but in the process the proof term (which has quadratic tree size) will be traversed, and without sufficiently precise caching to avoid scaling in the tree size.

Note: you need to bump https://github.com/leanprover-community/lean/blob/master/src/library/abstract_context_cache.cpp#L19 for the larger examples.
